### ISSUE TITLE
Add Rigidbody.GetPointVelocity(Vector3 point)

### ISSUE THREAD

**Issue by kodkuce** [godotengine/godot-proposals#773]
**Describe the project you are working on:**
Working on a kart game

**Describe the problem or limitation you are having in your project:**
I need to get velocity of rigidbody at point, something both UE4 and Unity have
GetUnrealWorldVelocityAtPoint(), Rigidbody.GetPointVelocity..

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Problem is i duno math to do it myself :)

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Hmm it would be a function that would return the value

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
I think its just an math/physics thingy so possible someone can write that function themself
but hmm i think that should be inbuild function

**Is there a reason why this should be core and not an add-on in the asset library?:**
Cuz its physics related and probably people will use it a lot, dont think it is exposed in
both UE4 and Unity for no reason

**Comment by Calinou** [comment]
@pouleyKetchoupp Is this easy enough to do with a script, or should it be a built-in feature?

**Comment by pouleyKetchoupp** [comment]
It's just one line of script, but with this kind of math, I agree it would be nice to add to the API so users don't have to figure out the right calculation, including the center of mass, etc.

I would put it in `PhysicsBody` node so it can be accessed for any type of body, and it could be something like:
`get_velocity_at_local_point(Vector3 local_point)`: velocity at a position relative to the body's center of mass
`get_velocity_at_world_point(Vector3 world_point)`: velocity at a given world position

**Comment by 1player** [comment]
This can probably be closed, as it already exists: https://docs.godotengine.org/en/stable/classes/class_physicsdirectbodystate3d.html#class-physicsdirectbodystate3d-method-get-velocity-at-local-position

This is accessible using RigidBody's _integrate_forces callback.

**Comment by addmix** [comment]
I see no reason why it shouldn't be accessible from any scope. For that reason, I think this proposal should stay open.

The implementation is pretty simple, so it should make for a good first issue.

Reference implementation: https://github.com/addmix/godot_utils/blob/b2d9e94d7b5484677a1b3c9af44c10d5a8b7702e/globals/physics_3d_utils.gd#L10

**Comment by EdwardChanCH** [comment]
I can try opening a PR for this.

Existing code:
- https://github.com/godotengine/godot/blob/c939bf3791ce40ff70e0ee29f06486da1ebb6a84/modules/godot_physics_3d/godot_body_direct_state_3d.cpp#L98-L100
- https://github.com/godotengine/godot/blob/c939bf3791ce40ff70e0ee29f06486da1ebb6a84/modules/godot_physics_3d/godot_body_3d.h#L323-L325
- https://github.com/godotengine/godot/blob/c939bf3791ce40ff70e0ee29f06486da1ebb6a84/modules/jolt_physics/objects/jolt_physics_direct_body_state_3d.cpp#L101-L103
- https://github.com/godotengine/godot/blob/c939bf3791ce40ff70e0ee29f06486da1ebb6a84/modules/jolt_physics/objects/jolt_body_3d.cpp#L787-L798
- (and their 2D counterparts)

**Comment by addmix** [comment]
> I can try opening a PR for this.
> 
> Existing code:
> 
> * https://github.com/godotengine/godot/blob/c939bf3791ce40ff70e0ee29f06486da1ebb6a84/modules/godot_physics_3d/godot_body_direct_state_3d.cpp#L98-L100
> * https://github.com/godotengine/godot/blob/c939bf3791ce40ff70e0ee29f06486da1ebb6a84/modules/godot_physics_3d/godot_body_3d.h#L323-L325
> * https://github.com/godotengine/godot/blob/c939bf3791ce40ff70e0ee29f06486da1ebb6a84/modules/jolt_physics/objects/jolt_physics_direct_body_state_3d.cpp#L101-L103
> * https://github.com/godotengine/godot/blob/c939bf3791ce40ff70e0ee29f06486da1ebb6a84/modules/jolt_physics/objects/jolt_body_3d.cpp#L787-L798
> * (and their 2D counterparts)

I will say, I think the `get_velocity_in_local_point()` naming is confusing, as "in_local" might make it seem like it would be relative to the node's rotation, as well as not being consistent with Godot naming conventions, where `linear_velocity` is used to avoid ambiguity. 

- `get_linear_velocity_at_position()` may be a better name, or 
- `get_linear_velocity()` could be overloaded with an optional position parameter, similar to `apply_force()` and `apply_impulse()`, i.e.
`get_linear_velocity(const Vector3 &rel_pos = Vector3()) const {...}`. In this case, I think it would be best for this overload to be exposed to GDScript instead of the original, non-point-velocity function.

**Comment by EdwardChanCH** [comment]
> I will say, I think the get_velocity_in_local_point() naming is confusing, as "in_local" might make it seem like it would be relative to the node's rotation, as well as not being consistent with Godot naming conventions, where linear_velocity is used to avoid ambiguity...

Yeah, I agree that `GodotBody3D::get_velocity_in_local_point()` and `JoltBody3D::get_velocity_at_position()` are confusing (internal) names, but they also have (internal) wrapper functions `GodotPhysicsDirectBodyState3D::get_velocity_at_local_position()` and `JoltPhysicsDirectBodyState3D::get_velocity_at_local_position()`.

Only `PhysicsDirectBodyState3D.get_velocity_at_local_position()` is exposed in GDScript, I think.

`get_linear_velocity_at_position()` may not be as clear, since the result is the sum of both `linear_velocity` and `angular_velocity`.

`get_linear_velocity()` is a getter for `linear_velocity`; I don't think changing a getter is a good idea. A dedicated method is more visible in the docs.

For consistency, it may be better to follow the current trend and add `RigidBody3D.get_velocity_at_local_position()`.

Optionally, we can add a helper method `RigidBody3D.get_velocity_at_position()` that takes a `global_position` directly.

**Comment by addmix** [comment]
> get_linear_velocity_at_position() may not be as clear, since the result is the sum of both linear_velocity and angular_velocity.

That's actually a very good point that I hadn't though of.
